chapter_generative-adversarial-networks/dcgan.md

```diff
@@ -229,19 +376,93 @@ def train(net_D, net_G, data_iter, num_epochs, lr, latent_dim,
           f'{metric[2] / timer.stop():.1f} examples/sec on {str(device)}')
 ```
 
+```{.python .input}
+#@tab pytorch
+def update_D(X, Z, net_D, net_G, loss, trainer_D):
+    """Update discriminator."""
+    batch_size = X.shape[0]
+    ones = torch.ones((batch_size, 1, 1, 1), device=X.device)
+    zeros = torch.zeros((batch_size, 1, 1, 1), device=X.device)
+    trainer_D.zero_grad()
+    real_Y = net_D(X)
+    fake_X = net_G(Z)
+    # Do not need to compute gradient for `net_G`, detach it from
+    # computing gradients.
+    fake_Y = net_D(fake_X.detach())
+    loss_D = (loss(real_Y, ones) + loss(fake_Y, zeros)) / 2
+    loss_D.backward()
+    trainer_D.step()
+    return loss_D
+
+def update_G(Z, net_D, net_G, loss, trainer_G):
+    """Update generator."""
+    batch_size = Z.shape[0]
+    ones = torch.ones((batch_size,1,1,1), device=Z.device)
+    trainer_G.zero_grad()
+    # We could reuse `fake_X` from `update_D` to save computation
+    fake_X = net_G(Z)
+    # Recomputing `fake_Y` is needed since `net_D` is changed
+    fake_Y = net_D(fake_X)
+    loss_G = loss(fake_Y, ones)
+    loss_G.backward()
+    trainer_G.step()
+    return loss_G
```

Comment:
Use d2l.update... like the mxnet version

Reply:
torch.ones((batch_size, 1, 1, 1)
I can't. The num of dim isn't same.

Reply:
.ones((batch_size,))
mxnet can add dims auto.
But pytorch need to require dim advancely.

Reply:
Look GAN's d2l.update: ![image](https://user-images.githubusercontent.com/25657787/91970339-17bf5780-ed4a-11ea-9cde-da040c0e6bd2.png)

I have tried, but failed.